src/main/drivers/dma.c

```diff
@@ -24,60 +24,99 @@
+
+#define DEFINE_DMA_IRQ_HANDLER(d, c, i) \
+    void DMA ## d ## _Channel ## c ## _IRQHandler(void) {\
+        DMA_IRQHandler(i);\
+    } \
+    struct dummy
+
 
 /*
- * DMA handlers for DMA resources that are shared between different features depending on run-time configuration.
+ * DMA descriptors.
  */
-static dmaHandlers_t dmaHandlers;
+static dmaChannelDescriptor_t dmaDescriptors[] = {
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel1,  0, DMA1_Channel1_IRQn, RCC_AHBPeriph_DMA1),
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel2,  4, DMA1_Channel2_IRQn, RCC_AHBPeriph_DMA1),
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel3,  8, DMA1_Channel3_IRQn, RCC_AHBPeriph_DMA1),
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel4, 12, DMA1_Channel4_IRQn, RCC_AHBPeriph_DMA1),
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel5, 16, DMA1_Channel5_IRQn, RCC_AHBPeriph_DMA1),
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel6, 20, DMA1_Channel6_IRQn, RCC_AHBPeriph_DMA1),
+    DEFINE_DMA_CHANNEL(DMA1, DMA1_Channel7, 24, DMA1_Channel7_IRQn, RCC_AHBPeriph_DMA1),
+#if defined(STM32F3) || defined(STM32F10X_CL)
+    DEFINE_DMA_CHANNEL(DMA2, DMA2_Channel1,  0, DMA2_Channel1_IRQn, RCC_AHBPeriph_DMA2),
+    DEFINE_DMA_CHANNEL(DMA2, DMA2_Channel2,  4, DMA2_Channel2_IRQn, RCC_AHBPeriph_DMA2),
+    DEFINE_DMA_CHANNEL(DMA2, DMA2_Channel3,  8, DMA2_Channel3_IRQn, RCC_AHBPeriph_DMA2),
+    DEFINE_DMA_CHANNEL(DMA2, DMA2_Channel4, 12, DMA2_Channel4_IRQn, RCC_AHBPeriph_DMA2),
+    DEFINE_DMA_CHANNEL(DMA2, DMA2_Channel5, 16, DMA2_Channel5_IRQn, RCC_AHBPeriph_DMA2),
+#endif
+};
 
-void dmaNoOpHandler(DMA_Channel_TypeDef *channel)
-{
-    UNUSED(channel);
-}
 
-void DMA1_Channel2_IRQHandler(void)
+void DMA_IRQHandler(int id)
 {
-    dmaHandlers.dma1Channel2IRQHandler(DMA1_Channel2);
+    dmaCallbackHandler_t *h = dmaDescriptors[id].handler;
+    while (h) {
+        h->fn(&dmaDescriptors[id], h);
+        h = h->next;
+    }
 }
 
-void DMA1_Channel3_IRQHandler(void)
-{
-    dmaHandlers.dma1Channel3IRQHandler(DMA1_Channel3);
-}
 
-void DMA1_Channel6_IRQHandler(void)
-{
-    dmaHandlers.dma1Channel6IRQHandler(DMA1_Channel6);
-}
+/*
+ * DMA IRQ Handlers
+ */
+DEFINE_DMA_IRQ_HANDLER(1, 1, DMA1_CH1_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(1, 2, DMA1_CH2_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(1, 3, DMA1_CH3_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(1, 4, DMA1_CH4_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(1, 5, DMA1_CH5_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(1, 6, DMA1_CH6_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(1, 7, DMA1_CH7_HANDLER);
+
+#if defined(STM32F3) || defined(STM32F10X_CL)
+DEFINE_DMA_IRQ_HANDLER(2, 1, DMA2_CH1_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(2, 2, DMA2_CH2_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(2, 3, DMA2_CH3_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(2, 4, DMA2_CH4_HANDLER);
+DEFINE_DMA_IRQ_HANDLER(2, 5, DMA2_CH5_HANDLER);
+#endif
 
-void DMA1_Channel7_IRQHandler(void)
+
+void dmaInit(void)
 {
-    dmaHandlers.dma1Channel7IRQHandler(DMA1_Channel7);
+    // TODO: Do we need this?
```

Comment:
In future, DMA resource tracing may be implemented. And some other initialization may appear.
And only constant part of handler descriptors may be stored in flash, initializing rest of variables here in loop